### PR TITLE
Add rules for decorrelation of aggregations

### PR DESCRIPTION
**Pull request by kasiafi**
Fixes https://github.com/prestosql/presto/issues/5904
Fixes https://github.com/prestosql/presto/issues/6145

### CODE REVIEW

File: presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/TransformCorrelatedAggregationToJoin.java [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+        // append non-null symbol on source. It will be used to restore semantics of count(*) aggregations after LEFT join
+        if (nonNull.isPresent()) {
+            source = new ProjectNode(
+                    context.getIdAllocator().getNextId(),
+                    source,
+                    Assignments.builder()
+                            .putIdentities(source.getOutputSymbols())
+                            .put(nonNull.get(), TRUE_LITERAL)
+                            .build());
+        }
```
[the remark this answers is not in the document]
Comment (martint) [reply presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/TransformCorrelatedAggregationToJoin.java:98]: Indeed, at least for TransformCorrelatedDistinctAggregationWithoutProjection, it's a lot clearer what's going on if you inline this method and the calls to appendAggregations. They seem like unnecessary abstractions that don't add much value.

**Comment by kasiafi** [comment]
Inlined methods.

File: presto-main/src/test/java/io/prestosql/sql/query/TestCorrelatedAggregation.java [hunk 2]
```diff
@@ -0,0 +1,352 @@
...
+ */
+package io.prestosql.sql.query;
+
+import org.testng.annotations.AfterClass;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.Test;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class TestCorrelatedAggregation
```
Comment (martint) [inline presto-main/src/test/java/io/prestosql/sql/query/TestCorrelatedAggregation.java:22]: Add some tests involving `ORDER BY` within aggregations and `FILTER (WHERE ...)`

File: presto-main/src/test/java/io/prestosql/sql/query/TestCorrelatedAggregation.java [hunk 3]
```diff
@@ -0,0 +1,352 @@
...
+    public void teardown()
+    {
+        assertions.close();
+        assertions = null;
+    }
+
+    @Test
+    public void testGlobalDistinctAggregation()
+    {
+        assertThat(assertions.query("SELECT * FROM " +
```
Comment (martint) [inline presto-main/src/test/java/io/prestosql/sql/query/TestCorrelatedAggregation.java:42]: Use lower-case for table names. `T` -> `t`

**Pull request merged**